handbook/python-dataplugins/python-dataplugins.md

```diff
@@ -0,0 +1,264 @@
+               "values": [1, 2, 3, 4]
+            }
+            ]
+      }]
+   }
+
+   return {Path(file_path).stem: tdm_tree}
+```
+
+
+</p>
+</details>
+
+Use the file parameter to access your file using text, csv or binary readers. The data has to be filled into a python dictionary. It represents the [structure of tdm/tdms files](https://www.ni.com/en-us/support/documentation/supplemental/06/the-ni-tdms-file-format.html) that consist of one root, 0...m groups and 0...n channels:
+
+<img alt="TDM structure with file, groups and channels" src="../img/pydp-tdm_structure.png?raw=true" width="500"><br>
+
+<details>
+<summary>Example dictionary</summary>
+<p>
+
+```python 
+self.tdm_tree = {
+   "author": "National Instruments",
+   "description": "Example file",
+   "groups": [{
+         "name": "Example",
+         "description": "The first group",
+         "time": datetime.datetime(2020, 2, 11, 15, 31, 59, 342380),
+         "channels": [{
+            "name": "Channel_0",
+            "description": "",
+            "values": [1.2, 1.3, 1.4],
+            "info": "Time in seconds",
+            "type": "DataTypeChnFloat64"
+         }, {
+            "name": "Channel_1",
+            "description": "",
+            "values": [10, 11, 12],
+            "unit_string": "km/h",
+            "type": "DataTypeChnFloat64"
+         }]
+   }]
+}
+
+file_path = os.path.realpath(parameter["file"])
+return {Path(file_path).stem: self.tdm_tree}
+```
+</p>
+</details>
+
+<details>
+<summary>Dictionary Schema</summary>
+<p>
+
+```python
+import datetime
+from schema import And, Schema
+
+Schema({
+      Optional('author'): str,
+      Optional('description'): str,
+      'groups': [{
+         'name': str,
+         Optional('description'): str,
+         Optional('time'): datetime.datetime,
+         'channels': [{
+            'name': str,
+            Optional('description'): str,
+            'values': list,
+            Optional('unit_string'): str,
+            'type': And(str, lambda s: s in (
+               'DataTypeChnFloat32',
+               'DataTypeChnFloat64',
+               'DataTypeChnString',
+               'DataTypeChnDate',
+               'DataTypeChnUInt8',
+               'DataTypeChnInt16',
+               'DataTypeChnInt32',
+               'DataTypeChnInt64'))
+            }]
+      }]}, ignore_extra_keys=True)
+```
+
+</p>
+<p>All further extra keys will show up as custom properties in DIAdem, Labview or SystemLink DataFinder.</p>
+</details>
+
+See full example: [csv-read-with-direct-loading](../../examples/python-dataplugin-examples/csv-read-with-direct-loading/Readme)
+
+## Callback Loading
+When handling big data sets, it can make sense to not load all data at once. Instead, the DataPlugin should return only the values that are requested by the applications. We skip the costs that are needed to fill the value-arrays inside the `read_store` function - and just leave the array empty:

```

Comment:
I think this section could be more concise. I also don't understand the last sentence -- do people use the empty array below to ensure the DataPlugin only returns values the applications request? Or is that array the result of something?

Perhaps something like: 

'When handling a big data set, you might not want to load all data at the same time. To ensure the DataPlugin only returns values that applications request, [do x]'

Reply:
I tried to restructure the instructions. They are more clear now. Let me know what you think.